### ISSUE TITLE
Don't use buildscript.classpath configuration for KtLint runtime dependencies

### ISSUE THREAD

**Issue by mkobit** [jeremymailen/kotlinter-gradle#58]
Right now the documentation has:

```groovy
buildscript {
    configurations.classpath {
        resolutionStrategy { force 'com.github.shyiko:ktlint:0.19.0' }
    }
}

buildscript {
    dependencies {
        classpath files('libs/my-custom-ktlint-rules.jar')
        classpath 'org.other.ktlint:custom-rules:1.0'
    }
}
```

These dependencies are added as dependencies to the build script itself. These can potentially conflict with other transitive dependencies from other plugins and are also leaked in as compile/runtime dependencies of the build script. (Documentation at https://docs.gradle.org/current/userguide/tutorial_using_tasks.html#sec:build_script_external_dependencies)

Instead, it might be better to follow the Gradle convention of having a configuration specifically to group all Ktlint dependencies:

```groovy
configurations {
  ktlint
}
```

Which could then be used normally from the the `dependencies {}` block:

```
dependencies {
  ktlint(files(...))
  ktlint('somegroup:someid:1.0.0')
}
```

This is similar to what Gradle currently does with the FindBugs plugin (https://docs.gradle.org/4.8/userguide/findbugs_plugin.html), JaCoCo (https://docs.gradle.org/4.8/userguide/jacoco_plugin.html), CodeNarc (https://docs.gradle.org/4.8/userguide/codenarc_plugin.html), and other code quality plugins.

**Comment by jeremymailen** [comment]
Quite right, thank you for the observation @mkobit! This topic has been bouncing around in my mind for a while. I had worked on some code to do it, but it's a little bit complicated since it usually involves either a custom classloader or spawning a Java process. I want to maintain the tight incremental build feedback in-process and much of the code gradle uses to implement isolated dependency configuration for the above plugins resides in `internal` packages so shouldn't be reused.

I think it's all doable but would require some concentrated coding I haven't found time for yet. I'm open to contributions if someone has expertise implementing this!

Practically speaking I haven't (yet) heard anyone running into conflicts with `ktlint`'s dependencies -- except of course the kotlin compiler itself which I've had to adjust version compatibility for a few times. I'd had rationalized to myself that it would be odd to for someone to compile their code with a much different version of kotlin than they are using to check lint and formatting. Realistically this gets finicky over time though.

So to summarize, you are bringing up a very valid issue, it hasn't reached a level of urgency for me to address yet, but I'd be grateful if someone else could implement it elegantly while maintaining the plugin's functionality.

**Comment by mateuszkwiecinski** [comment]
Hey :)  I decided to give it a try and I tried to sketch something that runs `LintTask` using a classpath separated from the project buildscript. 
I created _something_ and [you can check it out](https://github.com/mateuszkwiecinski/kotlinter-gradle/compare/feature/lazy_sources_evaluation...mateuszkwiecinski:feature/use_custom_confguration_for_ktlint_dependencies?expand=1). However, I'm not creating the PR because that hasn't been fully tested nor I havent' understood what is the expected output from `Reporter` classes. What their additional purpose except of logging things to a file.
Another thing is that I'm fairly sure that my solution isn't the most efficient. I introduced additional filesystem access due to new Worker API restrictions (or simply because of lack of understanding of _Gradle_ framework). Also, I didn't have any idea **how to test the performance** - it would be great to have tests for that. As you said, it doesn't seem like a common issue - so loosing performance in favor of extreamly rare case doesn't sound like a good thing to do.
I'm writing this comment just to share my experiences.

>it usually involves either a custom classloader or spawning a Java process.

I noticed new Gradle Worker API (currently used `submit()` is already deprecated) eposes set of easily configurable isolation modes and hoped that would allow to separate classpaths. I used the [`classLoaderIsolation()`](https://guides.gradle.org/using-the-worker-api/#changing_the_isolation_mode) from Worker API, but it turned out: 
a) common services like `Logger` are not available, despite they should - Related issue: https://github.com/gradle/gradle/issues/2678 
b) When using any isolation mode - it is not possible to share any static object (like currently `ExecutionContextRepository` is) so the only thing that came to my mind was to _return_ data from the worker using a temporary file. 

So, assuming I didn't do any significant mistakes, to achieve classpath isolation using Workers API in an efficient way, there are 3 options:
1. `Kotlinter` would need to stop using Reporters and use its own implementation that would write to a file directly from the worker's `WorkAction`.
2. Hopefuly Workes will support returning data https://github.com/gradle/gradle/issues/8233 so the whole implementation would be simplified.
3. Add performance test and proove that my solution isn't that bad 😛

**Comment by jeremymailen** [comment]
Thank you for looking into it! Promising investigation. The rightly point out that our implementation of worker api is already deprecated. I haven't spent enough time looking into it, but when trying to understand the current state of the worker api before I couldn't tell if the Gradle team was done changing it or not -- it seemed to have some rough edges I was hoping would get ironed out.

Regarding the reports, there's the standard ktlint api for writing output artifacts and several flavors of them. The way the files are chunked for performance reasons right now, there might be multiple workers per sourceSet and therefore multiple workers writing to one output report file, so doing it directly from a worker without a concurrency safe output to it might cause trouble.

That said, I'd be willing to eventually lose multiple concurrent workers per sourceSet if we got classpath isolation in return. Maybe a little bit slower, but eliminate the hard to track down kotlin internal class conflicts.

**Comment by jeremymailen** [comment]
I made a serious examination of this approach here: https://github.com/jeremymailen/kotlinter-gradle/pull/154/commits/73a797aa2512b1ac436693544e8b019806fb213d

It work, however the performance suffers significantly and we run gradle out of metaspace sometimes. Perhaps the gradle worker api is not quite ready for this (in contrast to the ant runner used by the code quality plugins for the same thing).

For now I'm closing this. I think the rare compatibility caveats are worth the speed.
The code has been refactored to make it easy to add the feature should we change our mind.
